rust/yarp/src/lib.rs

```diff
@@ -0,0 +1,321 @@
+            Some(current)
+        }
+    }
+}
+
+/// A struct created by the `comments` method on `ParseResult`. It can be used
+/// to iterate over the comments in the parse result.
+pub struct Comments<'pr> {
+    comment: *mut yp_comment_t,
+    marker: PhantomData<&'pr yp_comment_t>
+}
+
+impl<'pr> Iterator for Comments<'pr> {
+    type Item = Comment<'pr>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match NonNull::new(self.comment) {
+            None => None,
+            Some(comment) => {
+                let current = Comment { comment, marker: PhantomData };
+                self.comment = unsafe { comment.as_ref().node.next.cast::<yp_comment_t>() };
+                Some(current)
+            }
+        }
+    }
+}
+
+/// The result of parsing a source string.
+pub struct ParseResult<'pr> {
+    source: &'pr [u8],
+    parser: NonNull<yp_parser_t>,
+    node: NonNull<yp_node_t>
+}
+
+impl<'pr> ParseResult<'pr> {
+    /// Returns the source string that was parsed.
+    #[must_use]
+    pub const fn source(&self) -> &'pr [u8] {
+        self.source
+    }
+
+    /// Returns a slice of the source string that was parsed using the given
+    /// location range.
+    #[must_use]
+    pub fn slice(&self, location: &Location<'pr>) -> &'pr [u8] {
+        let root = self.source.as_ptr().cast::<c_char>();
+
+        let start = usize::try_from(unsafe { location.start().offset_from(root) }).expect("start should point to memory after root");
+        let end = usize::try_from(unsafe { location.end().offset_from(root) }).expect("end should point to memory after root");
+
+        &self.source[start..end]
+    }
+
+    /// Returns an iterator that can be used to iterate over the errors in the
+    /// parse result.
+    #[must_use]
+    pub fn errors(&self) -> Diagnostics<'_> {
+        unsafe {
+            let list = &mut (*self.parser.as_ptr()).error_list;
+            Diagnostics { diagnostic: list.head.cast::<yp_diagnostic_t>(), marker: PhantomData }
+        }
+    }
+
+    /// Returns an iterator that can be used to iterate over the warnings in the
+    /// parse result.
+    #[must_use]
+    pub fn warnings(&self) -> Diagnostics<'_> {
+        unsafe {
+            let list = &mut (*self.parser.as_ptr()).warning_list;
+            Diagnostics { diagnostic: list.head.cast::<yp_diagnostic_t>(), marker: PhantomData }
+        }
+    }
+
+    /// Returns an iterator that can be used to iterate over the comments in the
+    /// parse result.
+    #[must_use]
+    pub fn comments(&self) -> Comments<'_> {
+        unsafe {
+            let list = &mut (*self.parser.as_ptr()).comment_list;
+            Comments { comment: list.head.cast::<yp_comment_t>(), marker: PhantomData }
+        }
+    }
+
+    /// Returns the root node of the parse result.
+    #[must_use]
+    pub fn node(&self) -> Node<'pr> {
+        Node::new(self.node.as_ptr())
+    }
```

Comment:
Returning a reference like that without the node being a field on `self` is not possible unless you're creating a repr transparent wrapper on an unsized field (like `BStr` does).

The way this is done here with a lifetime generic parameter on the struct is the idiomatic way to do this. Similar to `std::slice::Iter`.

Reply:
The reason I did this is because a lot of the nodes hold only references to the source. For example to get the name of a method call, you would usually go back to the source and slice out the `message_loc` field. In this way the nodes themselves don't have to hold onto strings, they only hold onto pointers.

Also, we're planning on our own arena allocator, which is why the parser has the be around when you call `yp_node_destroy`. So the lifetimes have to be tied together.

One thing I think we could do would be to provide a `to_owned` or something that would copy out the source information and give you everything you needed so that you could keep the nodes around.